### PR TITLE
Fix #7966/#7964: Exporter `visibleOnly`, `exportHeader` and `exportFooter`

### LINKED ISSUE #7964: DataExporter: Option to globally suppress table header/footer export
**Description**
In my application, I like to see footers in some of my UI tables, but never in the corresponding exports (XLSX, PDF and XML).

I'm always frustrated when I have to rewrite my custom exporters for each new major PrimeFaces release as each new version brings a more sophisticated API (Nonetheless I really like the work done there!).

I am aware of the `exportFooterValue` attribute in p:Column, but this is not overly helpful due to the sheer amount of columns in use.

**Describe the solution you would like**
I like to have boolean exportFooter and exportHeader attributes on the DataExporter element which globally disables footer and header export.

It would also be great if defaults are sensible here:

PDF: exportFooter='false' exportHeader='true'
XLSX/XSL: exportFooter='false' exportHeader='true'
XML: exportFooter='false' exportHeader='false'

### LINKED ISSUE #7966: DataExporter: Option to ignore columns which are hidden via ColumnToggler
**Describe the defect**
**Description**
Data exporters will always export all rendered, exportable columns but not columns excluded via ColumnToggler.
This is confusing for users who may want to tune their table content before exporting.

In the past, I have added checks to my custom exporters for column visibility (isVisible()). I'm always frustrated when I have to rewrite my custom exporters for each new major PrimeFaces release as each new version brings a more sophisticated API (Nonetheless I really like the work done there!).

**Reproducer**
Will follow.

**Environment:**
 - PF Version: _11.0_
 - JSF + version: _ALL_
 - Affected browsers: _ALL_

**To Reproduce**
1. Exclude columns via ColumnToggler
2. Export the table
3. Notice excluded columns appearing in the output

**Expected behavior**
Data exporters should omit columns which are hidden via ColumnToggler.
Perhaps this should be added via an additional attribute to be backwards compatible here.

### PR DESCRIPTION
**Pull request by melloware**
Do not export columns that are not visible.

### CODE REVIEW

**Comment by tuerker** [comment]
@melloware I do not appreciate this change since this changes fundamental semantics. we have a lot of tables where we have "complex structured" columns that are not exported while we are providing for such columns a non-visible but exportable counterpart. if now non-visible by default implies non-exportable, then all these tables will not work anymore. i am pretty sure that many other users relying on the semantics implemented so far. 

the column attributes visible and exportable should definitively be kept independent. many thanks in advance for considering this argument.

**Comment by melloware** [comment]
Great point @tuerker   let me see if I can make it configurable.

**Comment by melloware** [comment]
OK added new `visibleOnly` option which is false by default so its like current PF8/10/11 behavior.

**Comment by melloware** [comment]
@tandraschko I think I found a bug while testing this I would like to discuss when you have time.  Not sure if you are off on holiday.

**Comment by melloware** [comment]
Fixed all my issues!

**Comment by tandraschko** [comment]
hmmm
i dont like to increase max constructor params in checkstyle, maybe use a builder?

**Comment by melloware** [comment]
I was thinking the same thing. I didn't like it either I will fix.

**Comment by melloware** [comment]
Set checkstyle back to 10 params and switched to builder pattern

**Comment by tandraschko** [comment]
the Builder doesnt look what we did for our SortMeta/Filter and so on, everything else looks good for me

**Comment by melloware** [comment]
@tandraschko I followed the same pattern `ExportConfiguration.java` was using.  Should they both be modified to use the same builder pattern as SortMeta?

**Comment by tandraschko** [comment]
i think so
its a a normal pattern that "setBlub" is void and a builder is just called "blub"

**Comment by melloware** [comment]
OK updated both classes to use the common builder pattern.

File: primefaces/src/main/java/org/primefaces/component/export/DataExporter.java [hunk 1]
```diff
@@ -249,30 +254,159 @@ public void restoreState(FacesContext context, Object state) {
...
+        values[13] = exporter;
 
         return (values);
     }
+
+    public static DataExporterBuilder builder() {
+        return new DataExporterBuilder();
+    }
+
+    public static class DataExporterBuilder {
```
Comment (tandraschko) [inline primefaces/src/main/java/org/primefaces/component/export/DataExporter.java]: just call it Builder?

File: primefaces/src/main/java/org/primefaces/component/export/ExportConfiguration.java [hunk 2]
```diff
@@ -30,81 +30,170 @@
...
-        return outputFileName;
+    public ExportConfiguration() {
+        // NOOP
+    }
+
+    public static ExportConfigurationBuilder builder() {
+        return new ExportConfigurationBuilder();
+    }
+
+    public static class ExportConfigurationBuilder {
```
Comment (tandraschko) [inline primefaces/src/main/java/org/primefaces/component/export/ExportConfiguration.java]: also just "Builder"?

**Comment by melloware** [comment]
renamed both to builder

**Pull request merged**